elyra/pipeline/pipeline_definition.py

```diff
@@ -515,13 +580,14 @@ def propagate_pipeline_default_properties(self):
         kv_properties = PipelineDefinition.get_kv_properties()
         self.primary_pipeline.convert_kv_properties(kv_properties)
 
-        pipeline_default_properties = self.primary_pipeline.get_property(pipeline_constants.PIPELINE_DEFAULTS, {})
+        pipeline_default_properties = self.primary_pipeline.get_property(PIPELINE_DEFAULTS, {})
         for node in self.pipeline_nodes:
             if not Operation.is_generic_operation(node.op):
                 continue
 
-            # Convert any key-value list node properties to the KeyValueList type
-            node.convert_kv_properties(kv_properties)
+            # Convert any key-value list node properties to the KeyValueList type if not done already
+            if node.kv_properties_not_converted(kv_properties):
+                node.convert_kv_properties(kv_properties)
```

Comment:
Seems like `convert_kv_properties()` could have its loops structured like `kv_properties_not_converted()` and just do the conversion or exit if a `KeyValueList` instance is encountered and remove `kv_properties_not_converted()`.  It looks like any KVL instance implies the entire thing has been converted and since only those properties in `kv_properties` are checked, this could simplify things a tad - assuming I'm understanding the overall flow.